### ISSUE TITLE
verdaccio crashes; uncaught exception ETIMEDOUT 443

### ISSUE THREAD

**Issue by rambabusaravanan** [verdaccio/verdaccio#1283]
**Describe the bug**
I am able to publish private packages. Configured uplinks our own registry server and able to do npm install successfully. But verdaccio crashes at random cases, mostly after npm install is completed successfully.

```
REQUEST make request https://registry.npmjs.org/-/npm/v1/security/audits
 fatal--- uncaught exception, please report this
Error: connect ETIMEDOUT 104.16.17.35:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1191:14)
```

**To Reproduce**
simply crashes when unable to connect 443

**Configuration File (cat ~/.config/verdaccio/config.yaml)** 
The only change I made is our own registry as uplinks
```yml
# a list of other known repositories we can talk to
uplinks:
  npmjs:
    url: https://ourserver.com:8443/nexus/repository/npm-all/
    strict_ssl: false
    ca: null
```

**Debugging output**
```
$ cat .npmrc 
registry=http://localhost:4873/
$ npm config get registry
http://localhost:4873/
$ verdaccio --version
3.11.6
$ node --version
v8.14.0
$ npm --version
6.9.0
```

**Additional context**
Not all the https calls are allowed in our organization. https://registry.npmjs.org is not allowed by default

**Comment by MartinLoeper** [comment]
Hey, a similar instance of the issue here!
I am using a private verdaccio repository as uplink and use verdaccio as a local registry on developers machine: 
`Verdaccio (running on Developer Machines) --> AWS ALB --> Verdaccio (running on AWS ECS)`

Most of the requests are working fine, but some of the requests are failing. I am not sure if they are failing randomly (I attached the logs). It seems they are failing after large packages such as the `aws-sdk` are requested...

I get a bunch of the errors looking like the following;
```
Error: connect EHOSTUNREACH 52.29.138.8:443
verdaccio    |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
verdaccio    |  http <-- 500, user: null(172.28.0.1), req: 'GET /common-tags/-/common-tags-1.8.0.tgz', error: internal server error
```

Once the first request failed with EHOSTUNREACH, all the following requests are also failing. Is there a known bug with upstream request handling (to authenticated private registries) in Verdaccio or so? This is driving me crazy for some hours now since I am not able to spot the error. =(

### Why is this an issue?
Builds on developer machines which use the local verdaccio cache fail with errors which look like the following:
```
npm ERR! code E404
npm ERR! 404 Not Found: agenda@2.0.2
```

They fail with random packages throwing a 404.
If I use `npm install` directly and bypassing the local Verdaccio instance (thus connecting to AWS Verdaccio directly), the installation works fine.

[config.txt](https://github.com/verdaccio/verdaccio/files/3187319/config.txt)
[docker-compose.txt](https://github.com/verdaccio/verdaccio/files/3187321/docker-compose.txt)

**Comment by juanpicado** [comment]
@MartinLoeper perhaps you need to increased the body size https://verdaccio.org/docs/en/configuration#max-body-size which by default is `10mb`

**Comment by MartinLoeper** [comment]
@juanpicado thanks for the hint. Unfortunately this does not change anything. I think the packages we are not exceeding the default limit either =(

I am wondering whether this is a strange bug in the underlying Node/Docker/Linux network stack, e.g. DNS resolution or so... 
If I knew the tooling to investigate this, I would look into it. Any idea how to debug this issue?

**Comment by juanpicado** [comment]
Well, I have this guide https://github.com/verdaccio/verdaccio/wiki/Running-and-Debugging-tests#unpliks-debug but is focused on unit testing..

But if you start verdaccio as:

````
NODE_DEBUG=request verdaccio
```
You will be able to inspect the request done to the uplink as is described in that guide. Perhaps that helps.

**Comment by MartinLoeper** [comment]
@juanpicado Sounds tough, but I will look into it and try my best. Thanks =)

**Comment by piyushkantm** [comment]
I am facing similar issue and also using S3 plugin to use the files on S3.

I think the issue is that due to some reason file is not present on S3 bucket but verdaccio thinks it is present. (Maybe the file was not uploaded to S3 because of *max-body-size*)

Is there any way to clear verdaccio data? I am trying to clear verdaccio data and then delete the lib directory in S3

Here is my server log: 

```
ubuntu@ip-172-31-33-48:~/.config/verdaccio$ NODE_DEBUG=request verdaccio
 warn --- config file  - /home/ubuntu/.config/verdaccio/config.yaml
 warn --- Plugin successfully loaded: s3-storage
 warn --- Plugin successfully loaded: htpasswd
 warn --- Plugin successfully loaded: audit
 warn --- http address - http://0.0.0.0:4873/ - verdaccio/3.0.0-beta.12
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
REQUEST { url: 'https://registry.npmjs.org/typescript',
  method: 'GET',
  headers:
   { 'If-None-Match': 'W/"aa59eaccd44a23386a5becbcafb7b21a"',
     Accept: 'application/json;',
     'Accept-Encoding': 'gzip',
     'User-Agent': 'npm (verdaccio/3.0.0-beta.12)',
     Via: '1.1 913bc20ad389 (Verdaccio)' },
  encoding: null,
  gzip: true,
  timeout: 30000,
  strictSSL: true,
  callback: [Function] }
REQUEST make request https://registry.npmjs.org/typescript
REQUEST onRequestResponse https://registry.npmjs.org/typescript 304 { date: 'Tue, 15 Oct 2019 08:07:40 GMT',
  connection: 'close',
  'set-cookie':
   [ '__cfduid=d1409dcd41b00d43643301647372f67221571126860; expires=Wed, 14-Oct-20 08:07:40 GMT; path=/; domain=.npmjs.org; HttpOnly' ],
  'cf-cache-status': 'HIT',
  'cache-control': 'max-age=300',
  'cf-ray': '52604d00f88823b8-IAD',
  age: '6327',
  etag: '"aa59eaccd44a23386a5becbcafb7b21a"',
  'expect-ct':
   'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"',
  'last-modified': 'Tue, 15 Oct 2019 06:17:07 GMT',
  vary: 'accept-encoding, accept',
  'x-amz-meta-rev': '1755-2728cfa68cedb10fc20df5b5a4982af2',
  server: 'cloudflare' }
 http --> 304, req: 'GET https://registry.npmjs.org/typescript' (streaming)
REQUEST reading response's body
REQUEST finish init function https://registry.npmjs.org/typescript
REQUEST response end https://registry.npmjs.org/typescript 304 { date: 'Tue, 15 Oct 2019 08:07:40 GMT',
  connection: 'close',
  'set-cookie':
   [ '__cfduid=d1409dcd41b00d43643301647372f67221571126860; expires=Wed, 14-Oct-20 08:07:40 GMT; path=/; domain=.npmjs.org; HttpOnly' ],
  'cf-cache-status': 'HIT',
  'cache-control': 'max-age=300',
  'cf-ray': '52604d00f88823b8-IAD',
  age: '6327',
  etag: '"aa59eaccd44a23386a5becbcafb7b21a"',
  'expect-ct':
   'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"',
  'last-modified': 'Tue, 15 Oct 2019 06:17:07 GMT',
  vary: 'accept-encoding, accept',
  'x-amz-meta-rev': '1755-2728cfa68cedb10fc20df5b5a4982af2',
  server: 'cloudflare' }
REQUEST end event https://registry.npmjs.org/typescript
REQUEST emitting complete https://registry.npmjs.org/typescript
 http --> 304, req: 'GET https://registry.npmjs.org/typescript', bytes: 0/0
 http <-- 200, user: piyush(122.177.51.91 via 172.31.23.219), req: 'GET /typescript', bytes: 0/483701
 http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
REQUEST { url: 'https://registry.npmjs.org/-/npm/v1/security/audits',
  method: 'POST',
  req:
   IncomingMessage {
     _readableState:
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: BufferList { head: null, tail: null, length: 0 },
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: false,
        ended: true,
        endEmitted: true,
        reading: false,
        sync: false,
        needReadable: false,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        paused: false,
        emitClose: true,
        destroyed: false,
        defaultEncoding: 'utf8',
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: false,
     domain: null,
     _events:
      [Object: null prototype] {
        end: [Function: resetHeadersTimeoutOnReqEnd],
        data: [Function],
        close: [Function] },
     _eventsCount: 3,
     _maxListeners: undefined,
     socket:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [TCP],
        _parent: null,
        _host: null,
        _readableState: [ReadableState],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: true,
        allowHalfOpen: true,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Server],
        _server: [Server],
        timeout: 120000,
        parser: [HTTPParser],
        on: [Function: socketOnWrap],
        _paused: false,
        _httpMessage: [ServerResponse],
        _peername: [Object],
        [Symbol(asyncId)]: 6123,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]:
         Timeout {
           _called: false,
           _idleTimeout: 120000,
           _idlePrev: [TimersList],
           _idleNext: [TimersList],
           _idleStart: 33256,
           _onTimeout: [Function: bound ],
           _timerArgs: undefined,
           _repeat: null,
           _destroyed: false,
           [Symbol(unrefed)]: true,
           [Symbol(asyncId)]: 6124,
           [Symbol(triggerId)]: 6123 },
        [Symbol(kBytesRead)]: 0,
        [Symbol(kBytesWritten)]: 0 },
     connection:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [TCP],
        _parent: null,
        _host: null,
        _readableState: [ReadableState],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: true,
        allowHalfOpen: true,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Server],
        _server: [Server],
        timeout: 120000,
        parser: [HTTPParser],
        on: [Function: socketOnWrap],
        _paused: false,
        _httpMessage: [ServerResponse],
        _peername: [Object],
        [Symbol(asyncId)]: 6123,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]:
         Timeout {
           _called: false,
           _idleTimeout: 120000,
           _idlePrev: [TimersList],
           _idleNext: [TimersList],
           _idleStart: 33256,
           _onTimeout: [Function: bound ],
           _timerArgs: undefined,
           _repeat: null,
           _destroyed: false,
           [Symbol(unrefed)]: true,
           [Symbol(asyncId)]: 6124,
           [Symbol(triggerId)]: 6123 },
        [Symbol(kBytesRead)]: 0,
        [Symbol(kBytesWritten)]: 0 },
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     httpVersion: '1.1',
     complete: true,
     headers:
      { 'x-forwarded-for': '122.177.51.91',
        'x-forwarded-proto': 'https',
        'x-forwarded-port': '443',
        host: 'https://registry.npmjs.org/',
        'x-amzn-trace-id': 'Root=1-5da57e60-28fe62c0e9ef77409b377e40',
        'content-length': '289',
        'user-agent': 'npm/6.9.0 node/v10.16.3 darwin x64',
        'npm-in-ci': 'false',
        'npm-scope': '',
        'npm-session': '2c42cea3a4a1a220',
        referer: 'install typescript@3.2.2',
        authorization: 'Bearer OQrvVQUn6coTbnzTkKvlNS5xUzFjMhMmqGMwYkSc5fU=',
        'content-type': 'application/json',
        'content-encoding': 'gzip',
        accept: '*/*',
        'accept-encoding': 'gzip,deflate' },
     rawHeaders:
      [ 'X-Forwarded-For',
        '122.177.51.91',
        'X-Forwarded-Proto',
        'https',
        'X-Forwarded-Port',
        '443',
        'Host',
        'registry.website.com',
        'X-Amzn-Trace-Id',
        'Root=1-5da57e60-28fe62c0e9ef77409b377e40',
        'Content-Length',
        '289',
        'user-agent',
        'npm/6.9.0 node/v10.16.3 darwin x64',
        'npm-in-ci',
        'false',
        'npm-scope',
        '',
        'npm-session',
        '2c42cea3a4a1a220',
        'referer',
        'install typescript@3.2.2',
        'authorization',
        'Bearer OQrvVQUn6coTbnzTkKvlNS5xUzFjMhMmqGMwYkSc5fU=',
        'content-type',
        'application/json',
        'content-encoding',
        'gzip',
        'accept',
        '*/*',
        'accept-encoding',
        'gzip,deflate' ],
     trailers: {},
     rawTrailers: [],
     aborted: false,
     upgrade: false,
     url: '/audits/quick',
     method: 'POST',
     statusCode: null,
     statusMessage: null,
     client:
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [TCP],
        _parent: null,
        _host: null,
        _readableState: [ReadableState],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [WritableState],
        writable: true,
        allowHalfOpen: true,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: [Server],
        _server: [Server],
        timeout: 120000,
        parser: [HTTPParser],
        on: [Function: socketOnWrap],
        _paused: false,
        _httpMessage: [ServerResponse],
        _peername: [Object],
        [Symbol(asyncId)]: 6123,
        [Symbol(lastWriteQueueSize)]: 0,
        [Symbol(timeout)]:
         Timeout {
           _called: false,
           _idleTimeout: 120000,
           _idlePrev: [TimersList],
           _idleNext: [TimersList],
           _idleStart: 33256,
           _onTimeout: [Function: bound ],
           _timerArgs: undefined,
           _repeat: null,
           _destroyed: false,
           [Symbol(unrefed)]: true,
           [Symbol(asyncId)]: 6124,
           [Symbol(triggerId)]: 6123 },
        [Symbol(kBytesRead)]: 0,
        [Symbol(kBytesWritten)]: 0 },
     _consuming: true,
     _dumped: false,
     next: [Function: next],
     baseUrl: '/-/npm/v1/security',
     originalUrl: '/-/npm/v1/security/audits/quick',
     _parsedUrl:
      Url {
        protocol: null,
        slashes: null,
        auth: null,
        host: null,
        port: null,
        hostname: null,
        hash: null,
        search: null,
        query: null,
        pathname: '/audits/quick',
        path: '/audits/quick',
        href: '/audits/quick',
        _raw: '/audits/quick' },
     params: {},
     query: {},
     res:
      ServerResponse {
        domain: null,
        _events: [Object],
        _eventsCount: 1,
        _maxListeners: undefined,
        output: [],
        outputEncodings: [],
        outputCallbacks: [],
        outputSize: 0,
        writable: true,
        _last: false,
        chunkedEncoding: false,
        shouldKeepAlive: true,
        useChunkedEncodingByDefault: true,
        sendDate: true,
        _removedConnection: false,
        _removedContLen: false,
        _removedTE: false,
        _contentLength: null,
        _hasBody: true,
        _trailer: '',
        finished: false,
        _headerSent: false,
        socket: [Socket],
        connection: [Socket],
        _header: null,
        _onPendingData: [Function: bound updateOutgoingData],
        _sent100: false,
        _expect_continue: false,
        req: [Circular],
        locals: [Object: null prototype] {},
        write: [Function: write],
        end: [Function: end],
        report_error: [Function],
        writeHead: [Function: writeHead],
        flush: [Function: flush],
        on: [Function: on],
        [Symbol(isCorked)]: false,
        [Symbol(outHeadersKey)]: [Object] },
     log:
      Logger {
        domain: null,
        _events: [Object: null prototype] {},
        _eventsCount: 0,
        _maxListeners: undefined,
        _level: 35,
        streams: [Array],
        serializers: [Object],
        src: false,
        fields: [Object] },
     body:
      { name: 'dln-s3-wrapper',
        version: '0.0.1',
        requires: [Object],
        dependencies: [Object],
        install: [Array],
        remove: [],
        metadata: [Object] },
     _body: true,
     route:
      Route { path: '/audits/quick', stack: [Array], methods: [Object] } },
  body:
   '{"name":"dln-s3-wrapper","version":"0.0.1","requires":{"typescript":"^3.2.2"},"dependencies":{"typescript":{"version":"3.2.2","integrity":"sha512-VCj5UiSyHBjwfYacmDuc/NOk4QQixbE+Wn7MFJuS0nRuPQbof132Pw4u53dm264O8LPc2MVsc7RJNml5szurkg==","dev":true}},"install":["typescript@3.2.2"],"remove":[],"metadata":{"npm_version":"6.9.0","node_version":"v10.16.3","platform":"darwin"}}',
  gzip: true,
  strictSSL: true,
  callback: [Function: requestCallback] }
REQUEST make request https://registry.npmjs.org/-/npm/v1/security/audits
REQUEST onRequestResponse https://registry.npmjs.org/-/npm/v1/security/audits 200 { date: 'Tue, 15 Oct 2019 08:08:01 GMT',
  'content-type': 'application/json; charset=utf-8',
  'transfer-encoding': 'chunked',
  connection: 'close',
  'set-cookie':
   [ '__cfduid=d85b52c8c83373307da801e23c8066d931571126880; expires=Wed, 14-Oct-20 08:08:00 GMT; path=/; domain=.npmjs.org; HttpOnly' ],
  'content-encoding': 'gzip',
  'cf-cache-status': 'DYNAMIC',
  'cf-ray': '52604d7ddb1b9fe8-IAD',
  'access-control-allow-credentials': 'true',
  'access-control-allow-origin': '*',
  'expect-ct':
   'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"',
  vary: 'Accept-Encoding',
  server: 'cloudflare' }
REQUEST reading response's body
REQUEST finish init function https://registry.npmjs.org/-/npm/v1/security/audits
REQUEST response end https://registry.npmjs.org/-/npm/v1/security/audits 200 { date: 'Tue, 15 Oct 2019 08:08:01 GMT',
  'content-type': 'application/json; charset=utf-8',
  'transfer-encoding': 'chunked',
  connection: 'close',
  'set-cookie':
   [ '__cfduid=d85b52c8c83373307da801e23c8066d931571126880; expires=Wed, 14-Oct-20 08:08:00 GMT; path=/; domain=.npmjs.org; HttpOnly' ],
  'content-encoding': 'gzip',
  'cf-cache-status': 'DYNAMIC',
  'cf-ray': '52604d7ddb1b9fe8-IAD',
  'access-control-allow-credentials': 'true',
  'access-control-allow-origin': '*',
  'expect-ct':
   'max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"',
  vary: 'Accept-Encoding',
  server: 'cloudflare' }
REQUEST end event https://registry.npmjs.org/-/npm/v1/security/audits
REQUEST has body https://registry.npmjs.org/-/npm/v1/security/audits 210
REQUEST emitting complete https://registry.npmjs.org/-/npm/v1/security/audits
 http <-- 200, user: undefined(122.177.51.91 via 172.31.23.219), req: 'POST /-/npm/v1/security/audits/quick', bytes: 289/210
 http <-- 200, user: piyush(122.177.51.91 via 172.31.23.219), req: 'GET /typescript/-/typescript-3.2.2.tgz', bytes: 0/7604389
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
{ http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
 http <-- 200, user: piyush(122.177.51.91 via 172.31.23.219), req: 'GET /typescript/-/typescript-3.2.2.tgz', bytes: 0/7604389
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.2.30), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.23.219), req: 'GET /', bytes: 0/640
 http <-- 200, user: undefined(172.31.43.160), req: 'GET /', bytes: 0/640
^C
ubuntu@ip-172-31-33-48:~/.config/verdaccio$ ^C
```

**Comment by DanielRuf** [comment]
@piyushkantm I see no actual error there. You should open a new issue with all relevant details, including the actual error message.

**Comment by DanielRuf** [comment]
Afaik there is some specific timeout for caching responses and marking an uplink as unreachable.

Does this issue still occur? In general this was a network issue at Cloudflare / npmjs.

**Comment by juanpicado** [comment]
> Is there any way to clear verdaccio data? I am trying to clear verdaccio data and then delete the lib directory in S3

No, it has to be done manually. 

@piyushkantm  As @DanielRuf suggest, let's create a new ticket, post relevant data and we follow up there all open questions. Topic here is different.
